### PR TITLE
Support DISC backend in torch_quant tool

### PR DESCRIPTION
**Pull request by deeply**

### CODE REVIEW

File: tools/torch_quant/torch_quant/observed_module.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+
+Linear = type('Linear', (nnqat.Linear, ), {
+              'from_float': classmethod(_from_float), 'forward': _linear_forward})
+
+OB_MODULE_MAPPING = {
+    nn.Conv1d: Conv1d,
+    nn.Conv2d: Conv2d,
+    nn.Conv3d: Conv3d,
+    nn.Linear: Linear,
```
Comment (linbinskn) [inline tools/torch_quant/torch_quant/observed_module.py:53]: Is it feasible to add func module like `torch.matmul` in current design?

File: tools/torch_quant/bert_ptq_demo.py [hunk 2]
```diff
@@ -41,27 +47,90 @@ def calibrate(model, dataloader, device):
...
+    print('load model...')
+    model = AutoModelForSequenceClassification.from_pretrained(
+        MODEL_ID, torchscript=True)
+    # model.config.problem_type is set in first forward calling, and is required for fx tracing
+    # we call calibrate() here only as a dummy forward calling.
+    calibrate(model, calib_dataloader, device)
+
+    tracer = HFTracerWrapper(
+        ["input_ids", "attention_mask", "token_type_ids"])
+    quantizer = Quantizer(tracer=tracer, backend=Backend.DISC)
```
Comment (linbinskn) [inline tools/torch_quant/bert_ptq_demo.py:107]: Is there any interface for us to set the quant module list?

File: tools/torch_quant/torch_quant/observed_module.py (hunk 1 above)
Comment (deeply) [reply tools/torch_quant/torch_quant/observed_module.py:53]: I think supporting quantizable functions like torch.matmul, is similar with supporting quantizable modules like nn.Linear. Need to implement it with a new "graph modification pass" though. Similar logic like `quantizable_module_to_observed` : find the call_function node in fx graph, and insert observer after weight and bias.

File: tools/torch_quant/bert_ptq_demo.py (hunk 2 above)
Comment (deeply) [reply tools/torch_quant/bert_ptq_demo.py:107]: There is a `module_filter` argument for `Quantizer` for this purpose. Not implemented yet 😂

File: tools/torch_quant/torch_quant/graph.py [hunk 3]
```diff
@@ -162,6 +165,34 @@ def insert_act_observer(ctx: GraphModContext) -> None:
     ctx.gm.recompile()
 
 
+def quantizable_module_to_observed(ctx: GraphModContext) -> None:
+    """
+    Replace quantizable modules with observed version.
+
+    For quantizable modules (e.g. Conv2D, Linear), weight is stored and used
+    internally, instead of as a sperated node in fx graph. So we need to
```
Comment (chenbohua3) [inline tools/torch_quant/torch_quant/graph.py:173]: typo: `sperated`

File: tools/torch_quant/torch_quant/observed_module.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    new_mod.bias_ob = bias_ob
+    delattr(new_mod, 'weight_fake_quant')
+    return new_mod
+
+
+def _conv_type(name: str, base: Type) -> Type:
+    def forward(self, input):
+        return self._conv_forward(input, self.w_ob(self.weight), self.bias_ob(self.bias))
+
+    return type(name, (base,), {'from_float': classmethod(_from_float), 'forward': forward})
```
Comment (chenbohua3) [inline tools/torch_quant/torch_quant/observed_module.py:34]: fancy:)

File: tools/torch_quant/torch_quant/quantizer.py [hunk 5]
```diff
@@ -27,16 +31,43 @@ class Backend(Enum):
     FBGEMM = 2
 
 
+DEFAULT_ACT_OB_CTR: Dict[Backend, Callable[..., Observer]] = {
```
Comment (chenbohua3) [inline tools/torch_quant/torch_quant/quantizer.py:34]: 1. `qscheme` is not set which will lead the `symmetric` of `MinMaxObserver` always `False` and further leads the wrong scales and zero_points.
2. Not a change introduced by this pull request but will affect accuracy, [here](https://github.com/alibaba/BladeDISC/blob/6c270d03d2521acb3a999f7d9dc8ad55efacfffb/tools/torch_quant/torch_quant/observer.py#L93), the zero point should to int data type.

File: tools/torch_quant/torch_quant/observer.py [hunk 6]
```diff
@@ -73,7 +74,7 @@ def __init__(self, dtype: torch.dtype = torch.qint8,
         self.register_buffer("min_val", torch.tensor(float("inf")))
         self.register_buffer("max_val", torch.tensor(float("-inf")))
         self.register_buffer("scale", torch.tensor(1.))
-        self.register_buffer("zero_point", torch.tensor(0.))
+        self.register_buffer("zero_point", torch.tensor(0, dtype=torch.int32))
```
Comment (chenbohua3) [inline tools/torch_quant/torch_quant/observer.py:77]: Should we support lower versions of torch. In earlier versions of torch, the zp is of `torch.long` dtype

File: tools/torch_quant/torch_quant/quantizer.py (hunk 5 above)
Comment (chenbohua3) [reply tools/torch_quant/torch_quant/quantizer.py:34]: And better to add some uts to check the scale & zero_point after the calibration step

File: tools/torch_quant/bert_ptq_demo.py [hunk 7]
```diff
@@ -22,6 +26,8 @@
 
 logging.basicConfig(level=logging.DEBUG)
 
+MODEL_ID = 'bert-base-cased-finetuned-mrpc'
```
[the remark this answers is not in the document]
Comment (deeply) [reply tools/torch_quant/bert_ptq_demo.py:29]: It may take some time to evaluate the accuracy of demo model. I'll leave a TODO here to use a smaller model as demo.

File: tools/torch_quant/torch_quant/graph.py (hunk 3 above)
Comment (deeply) [reply tools/torch_quant/torch_quant/graph.py]: fixed

File: tools/torch_quant/torch_quant/quantizer.py (hunk 5 above)
Comment (deeply) [reply tools/torch_quant/torch_quant/quantizer.py:34]: > 

1. OK, qscheme is specified for each backend observer constructors in the updated commit.
2. zero point will be implicitly rounded when we copy it to a tensor with dtype=int32?

Comment (deeply) [reply tools/torch_quant/torch_quant/quantizer.py:34]: there is a very basic test at [test_min_max_observer.py](https://github.com/alibaba/BladeDISC/blob/6c270d03d2521acb3a999f7d9dc8ad55efacfffb/tools/torch_quant/tests/test_min_max_observer.py#L19). I'll add a todo for more tests.

File: tools/torch_quant/torch_quant/observer.py (hunk 6 above)
Comment (deeply) [reply tools/torch_quant/torch_quant/observer.py:77]: I can't make a decision right now. maybe we can discuss how to support old versions (and which versions) when some use cases come out 😂

**Comment by chenbohua3** [comment]
LGTM

**Review by chenbohua3 — APPROVED** [review]
